gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/src/AuthorizationService.js

```diff
@@ -100,125 +100,510 @@ var fluid = fluid || require("infusion");
+                    }
+                });
+                return unauthorizedClients;
+            };
+            promiseTogo = fluid.promise.map(allClientsPromise, mapper);
+        }
+        return promiseTogo;
+    };
+    // ==== End of getUnauthorizedClientsForGpiiToken()
+
+    /*
+     * @promiseTogo: Modified by the function with objects to be resolved or to fail
+     * @dataStore (Object): a data store instance
+     * @codeGenerator (Object): a code generator
+     * @clientId (String): an unique client id
+     * @return: none. The first argument of promiseTogo contains return values
+     */
+    gpii.oauth2.authorizationService.createClientCredentialsToken = function (promiseTogo, dataStore, codeGenerator, clientId) {
+        var accessToken = codeGenerator.generateAccessToken();
+        var addClientCredentialsTokenPromise = dataStore.addClientCredentialsToken({
+            clientId: clientId,
+            accessToken: accessToken,
+            allowAddPrefs: true
         });
 
-        return unauthorizedClients;
+        var mapper = function () {
+            // The created access token is resolved for promiseTogo eventually
+            return accessToken;
+        };
+        var accessTokenPromise = fluid.promise.map(addClientCredentialsTokenPromise, mapper);
+        fluid.promise.follow(accessTokenPromise, promiseTogo);
     };
 
     gpii.oauth2.authorizationService.grantClientCredentialsAccessToken = function (dataStore, codeGenerator, clientId, scope) {
-        // Record the credential client access token if we haven't already
-        var client = dataStore.findClientById(clientId);
-        if (!scope || scope.indexOf("add_preferences") === -1 || !client.allowAddPrefs) {
-            return false;
-        }
+        var clientPromise = dataStore.findClientById(clientId);
+        var clientCredentialsTokenPromise = dataStore.findClientCredentialsTokenByClientId(clientId);
 
-        var clientCredentialsToken = dataStore.findClientCredentialsTokenByClientId(clientId);
-        if (!clientCredentialsToken) {
-            var accessToken = codeGenerator.generateAccessToken();
-            clientCredentialsToken = dataStore.addClientCredentialsToken({
-                clientId: clientId,
-                accessToken: accessToken,
-                allowAddPrefs: true
-            });
-        }
+        // TODO: Update the usage of fluid.promise.sequence() once https://issues.fluidproject.org/browse/FLUID-5938 is resolved.
+        var sources = [clientPromise, clientCredentialsTokenPromise];
+        var promisesSequence = fluid.promise.sequence(sources);
 
-        return clientCredentialsToken.accessToken;
+        var promiseTogo = fluid.promise();
+
+        promisesSequence.then(function (responses) {
+            var client = responses[0];
+            var clientCredentialsToken = responses[1];
+
+            if (!scope || scope.indexOf("add_preferences") === -1 || !client.allowAddPrefs) {
```

Comment:
Why might the scope not be exactly "add_preferences"? Is there some scope (ha ha) for allowing it to hold multiple strings or scopes?